sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -3503,6 +3503,42 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       assert(SQLConf.get.getConf(SQLConf.CODEGEN_FALLBACK) === true)
     }
   }
+
+  test("Perform propagating empty relation after RewritePredicateSubquery") {
```

Comment:
end-to-end optimizer test is useful, but `SQLQuerySuite` may not be a good place to put it.

How about we create a new test suite `EndToEndSQLOptimizerSuite`?

Reply:
sgtm, done